app/api/v2/managers/operation_api_manager.py

```diff
@@ -1,16 +1,89 @@
+import uuid
+
 from app.api.v2.managers.base_api_manager import BaseApiManager
-from app.api.v2.responses import JsonHttpNotFound, JsonHttpForbidden
+from app.api.v2.responses import JsonHttpNotFound, JsonHttpForbidden, JsonHttpBadRequest
+from app.objects.secondclass.c_link import Link
+from app.objects.secondclass.c_executor import Executor
+from app.objects.c_ability import AbilitySchema
+from app.objects.c_agent import Agent
+from app.objects.secondclass.c_executor import ExecutorSchema
+from app.objects.c_operation import Operation
+from app.utility.base_world import BaseWorld
 
 
 class OperationApiManager(BaseApiManager):
-    def __init__(self, data_svc, file_svc):
-        super().__init__(data_svc=data_svc, file_svc=file_svc)
+    def __init__(self, services):
+        super().__init__(data_svc=services['data_svc'], file_svc=services['file_svc'])
+        self.services = services
 
     async def get_operation_report(self, operation_id: str, access: dict):
         operation = await self.get_operation_object(operation_id, access)
         report = await operation.report(file_svc=self._file_svc, data_svc=self._data_svc)
         return report
 
+    async def get_operation_links(self, operation_id: str, access: dict):
+        operation = await self.get_operation_object(operation_id, access)
+        links = [link.display for link in operation.chain]
+        return links
+
+    async def get_operation_link(self, operation_id: str, link_id: str, access: dict):
+        operation = await self.get_operation_object(operation_id, access)
+        link = self.search_operation_for_link(operation, link_id)
+        return link.display
+
+    async def update_operation_link(self, operation_id: str, link_id: str, link_data: dict, access: BaseWorld.Access):
+        operation = await self.get_operation_object(operation_id, access)
+        link = self.search_operation_for_link(operation, link_id)
+        if link.access not in access['access']:
+            raise JsonHttpForbidden(f'Cannot update link {link_id} due to insufficient permissions.')
+        if link.is_finished() or link.can_ignore():
+            raise JsonHttpForbidden(f'Cannot update a finished link: {link_id}')
+        if link_data.get('status'):
+            link.status = link_data.get('status')
+        if link_data.get('command'):
+            link.command = link_data.get('command')
+            command_str = self._decode_string(link_data.get('command'))
+            link.executor.command = command_str
+        return link.display
+
+    async def create_potential_link(self, operation_id: str, data: dict, access: BaseWorld.Access):
+        self.validate_link_data(data)
+        operation = await self.get_operation_object(operation_id, access)
+        link_id = data.get('id', str(uuid.uuid4()))
```

Comment:
the `apply_link` method in c_link.py will actually generate a new random id for itself (so it'll overwrite any previous IDs). So the user doesn't even need to specify a link ID. So we don't need to pull it from the data dict and check if it's in the operation or not

Reply:
Sorry, missed removing that before - also updated endpoint decorator so that id won't appear as an input field